### PR TITLE
Fix deadlock induced MacOS PW Pool collapse

### PR DESCRIPTION
**Pull request by cevich**
Every night a script runs to check and possibly update all the scripts in the repo.  When this happens, two important activities take place:

1. The script is restarted (presuming it's own code changed).
2. The container running nginx (for the usage graph) is restarted.

For unknown reasons, possibly due to a system update, a pasta (previously slirp4netns) sub-process spawned by podman is holding open the lock-file required by both the maintenance script and the (very important) `Cron.sh`.  This leads to a deadlock situation where the entire pool becomes unmanaged since `Cron.sh` can't run.

To prevent unchecked nefarious/unintended use, all workers automatically recycle themselves after some time should they become unmanaged. Therefore, without `Cron.sh` operating, the entire pool will eventually collapse.

Though complex, as a (hopefully) temporary fix, ensure all non-stdio FDs are closed (in a sub-shell) prior to restarting the container.

---

Tested this change manually: Using a modified copy of the script which simply obtains the lock and always calls `relaunch_web_container()`.  Examining a before/after `lsof` embedded in that function, there is no more leaked `Cron.sh` FD open by pasta.

### CODE REVIEW

**Comment by cevich** [comment]
@edsantiago @Luap99 when you have a moment, any major SNAFU's with this change?  Once merged I'll manually update it on the management VM and monitor to be double sure it's working.

P.S. I know most of this stuff is really ugly, it was all intended to be temporary :cry:   Paul suggested a refactor using systemd timers which seems like a good idea.

**Review by Luap99 — APPROVED** [review]
LGTM

File: mac_pw_pool/nightly_maintenance.sh [hunk 1]
```diff
@@ -22,9 +22,21 @@ if [[ ! -r "$CRONLOG" ]] || [[ ! -r "$CRONSCRIPT" ]] || [[ ! -d "../.git" ]]; th
 fi
 
 relaunch_web_container() {
-  # Assume code has changed, restart container w/ latest image
+  # Assume code change or image update, restart container.
   (
+    # Prevent podman and/or sub-processes from inheriting the lock FD.
+    # This would deadlock all future runs of this script or Cron.sh
+    # Can't use `flock --close ...` here because it "hangs" in this context.
+    for fd_path in /proc/self/fd/*; do
```
Comment (edsantiago) [inline mac_pw_pool/nightly_maintenance.sh:30]: `for fd in $(/bin/ls /proc/self/fd)` saves you the (useless) intermediate variable and subsequent `basename`. But LGTM anyway.

File: mac_pw_pool/nightly_maintenance.sh [hunk 2]
```diff
@@ -22,9 +22,21 @@ if [[ ! -r "$CRONLOG" ]] || [[ ! -r "$CRONSCRIPT" ]] || [[ ! -d "../.git" ]]; th
...
+        fd_nr=$(basename $fd_path)
+        [[ $fd_nr -ge 3 ]] || \
+            continue
```
Comment (edsantiago) [inline mac_pw_pool/nightly_maintenance.sh:33]: Why this usage? Why not a simple `if`? (Just curious. My brain does not process `test ||` as easily as it does `if/then`).

**Review by edsantiago — APPROVED** [review]

File: mac_pw_pool/nightly_maintenance.sh (hunk 2 above)
Comment (Luap99) [reply mac_pw_pool/nightly_maintenance.sh:33]: agreed I also find if else easier to read in scripts

Comment (cevich) [reply mac_pw_pool/nightly_maintenance.sh:33]: My brain has an easier time understanding this as a "filter", "only greater/equal to 3 may proceed".   If it were a more complex condition in a different context, I'd probably agree with you though :smile:

**Comment by cevich** [comment]
Force-push: Simplified loop as Ed suggested.  Tested w/ test-script.

**Comment by cevich** [comment]
Manually updated maintenance VM w/ this new code.

**Comment by Luap99** [comment]
Pasta now closes all additional fds on startup which should help others from running into such a problem in the future
https://passt.top/passt/commit/?id=09603cab28f9883baf1d7b48bdc102d6641dc300

But I still think it is best to keep this even once a updated pasta is available to avoid issues with other child processes that might face the same issue.

**Comment by cevich** [comment]
That's great news, thanks Paul for pursuing that.  I'm sure it will help someone else avoid running into similar issues.  Yes, I completely agree about keeping this in place.

**Pull request merged**